x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngine.java

```diff
@@ -111,9 +112,14 @@ private void preFlight(final Operation operation) {
     }
 
     @Override
-    public NoOpResult noOp(NoOp noOp) {
-        // TODO: Make sure we process NoOp once.
-        return super.noOp(noOp);
+    protected Optional<Exception> preFlightCheckForNoOp(NoOp noOp) throws IOException {
+        if (noOp.origin() == Operation.Origin.PRIMARY && hasBeenProcessedBefore(noOp)) {
+            // See the comment in #indexingStrategyForOperation for the explanation why we can safely skip this operation.
+            final OptionalLong existingTerm = lookupPrimaryTerm(noOp.seqNo());
```

Comment:
random thought - I wonder if we should load the operation under assertion code and check it's the same (this goes for all duplicate ops).

Reply:
Do you mean the existing operation should equal the processing operation except for the primary term?

Reply:
yes

Reply:
++. I'll make it in a follow-up after this PR.